File: MoppLib/MoppLib/MoppLibError.m
//
//  MoppLibError.m
//  MoppLib
//
/*
 * Copyright 2017 - 2024 Riigi Infosüsteemi Amet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#import "MoppLibError.h"
#import "MoppLibConstants.h"

NSString *const MoppLibErrorDomain = @"MoppLibError";

@implementation MoppLibError

+ (NSError *)readerNotFoundError {
  return [self error:moppLibErrorReaderNotFound withMessage:@"Reader is not commected to device."];
}

+ (NSError *)readerProcessFailedError {
  return [self error:moppLibErrorReaderProcessFailed withMessage:@"Reader process failed."];
}

+ (NSError *)readerSelectionCanceledError {
  return [self error:moppLibErrorReaderSelectionCanceled withMessage:@"User canceled reader selection."];
}

+ (NSError *)cardNotFoundError {
  return [self error:moppLibErrorCardNotFound withMessage:@"ID card could not be detected in reader."];
}
    
+ (NSError *)ldapResponseNotFoundError {
    return [self error:moppLibErrorLdapResponseNotFound withMessage:@"Ldap response is empty"];
}

+ (NSError *)duplicatedFilenameError {
    return [self error:moppLibErrorDuplicatedFilename withMessage:@"Filename already exists"];
}

+ (NSError *)cardVersionUnknownError {
  return [self error:moppLibErrorCardVersionUnknown withMessage:@"Card version could not be detected."];
}

+ (NSError *)wrongPinErrorWithRetryCount:(int)count {
  return [self error:moppLibErrorWrongPin userInfo:@{kMoppLibUserInfoRetryCount:[NSNumber numberWithInt:count]}];
}

+ (NSError *)generalError {
  return [self error:moppLibErrorGeneral withMessage:@"Could not complete action due to unknown error"];
}

+ (NSError *)pinBlockedError {
  return [self error:moppLibErrorPinBlocked userInfo:nil];
}

+ (NSError *)invalidPinError {
  return [self error:moppLibErrorInvalidPin withMessage:@"Invalid PIN"];
}

+ (NSError *)pinNotProvidedError {
  return [self error:moppLibErrorPinNotProvided withMessage:@"PIN was not provided while it was required."];
}

+ (NSError *)pinMatchesVerificationCodeError {
  return [self error:moppLibErrorPinMatchesVerificationCode withMessage:@"New PIN must be different from verification code."];
}

+ (NSError *)pinMatchesOldCodeError {
  return [self error:moppLibErrorPinMatchesOldCode withMessage:@"New PIN must be different from old PIN."];
}

+ (NSError *)incorrectPinLengthError {
  return [self error:moppLibErrorIncorrectPinLength withMessage:@"PIN length didn't pass validation. Make sure minimum and maximum length requirements are met."];
}

+ (NSError *)tooEasyPinError {
  return [self error:moppLibErrorPinTooEasy withMessage:@"New PIN code is too easy."];
}

+ (NSError *)pinContainsInvalidCharactersError {
  return [self error:moppLibErrorPinContainsInvalidCharacters withMessage:@"New PIN contains invalid characters."];
}

+ (NSError *)urlSessionCanceledError {
  return [self error:moppLibErrorUrlSessionCanceled withMessage:@"Url session canceled"];
}

+ (NSError *)xmlParsingError {
  return [self error:moppLibErrorXmlParsingError withMessage:@"XML parsing error."];
}

+ (NSError *)fileNameTooLongError {
    return [self error:moppLibErrorFileNameTooLong withMessage:@"File name is too long"];
}

+ (NSError *)noInternetConnectionError {
  return [self error:moppLibErrorNoInternetConnection withMessage:@"Internet connection not detected."];
}

+ (NSError *)sslHandshakeError {
  return [self error:moppLibErrorSslHandshakeFailed withMessage:@"Failed to create ssl connection with host."];
}

+ (NSError *)restrictedAPIError {
  return [self error:moppLibErrorRestrictedApi withMessage:@"This API method is not supported on third-party applications."];
}

+ (NSError *)tooManyRequests {
  return [self error:moppLibErrorTooManyRequests withMessage:@"digidoc-service-error-too-many-requests"];
}

+ (NSError *)ocspTimeSlotError {
  return [self error:moppLibErrorOCSPTimeSlot withMessage:@"Invalid OCSP time slot"];
}

+ (NSError *)invalidProxySettingsError {
  return [self error:moppLibErrorInvalidProxySettings withMessage:@"Invalid proxy settings"];
}

+ (NSError *)error:(NSUInteger)errorCode withMessage:(NSString *)message {
  return [self error:errorCode userInfo:@{NSLocalizedDescriptionKey : message}];
}

+ (NSError *)error:(NSUInteger)errorCode userInfo:(NSDictionary *)userInfo {
  NSError *newError = [[NSError alloc] initWithDomain:@"MoppLib" code:errorCode userInfo:userInfo];
  return newError;
}

@end

File: MoppLib/MoppLib/MoppLibError.swift
//
//  MoppLibError.swift
//  MoppLib
//
/*
 * Copyright 2017 - 2024 Riigi Infosüsteemi Amet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

import Foundation

/// MoppLibError class providing predefined NSError instances
@objcMembers
public class MoppLibError: NSObject {
    /// Error domain for MoppLib errors
    static let MoppLibErrorDomain = "MoppLibError"

    public static func readerNotFoundError() -> NSError {
        return error(code: .moppLibErrorReaderNotFound, message: "Reader is not connected to the device.")
    }

    public static func readerProcessFailedError() -> NSError {
        return error(code: .moppLibErrorReaderProcessFailed, message: "Reader process failed.")
    }

    public static func readerSelectionCanceledError() -> NSError {
        return error(code: .moppLibErrorReaderSelectionCanceled, message: "User canceled reader selection.")
    }

    public static func cardNotFoundError() -> NSError {
        return error(code: .moppLibErrorCardNotFound, message: "ID card could not be detected in reader.")
    }

    public static func cardVersionUnknownError() -> NSError {
        return error(code: .moppLibErrorCardVersionUnknown, message: "Card version could not be detected.")
    }

    public static func wrongPinError(withRetryCount count: Int) -> NSError {
        return error(code: .moppLibErrorWrongPin, userInfo: [kMoppLibUserInfoRetryCount: NSNumber(value: count)])
    }

    public static func generalError() -> NSError {
        return error(code: .moppLibErrorGeneral, message: "Could not complete action due to unknown error")
    }

    public static func invalidPinError() -> NSError {
        return error(code: .moppLibErrorInvalidPin, message: "Invalid PIN")
    }

    public static func pinBlockedError() -> NSError {
        return error(code: .moppLibErrorPinBlocked, userInfo: nil)
    }

    public static func pinMatchesVerificationCodeError() -> NSError {
        return error(code: .moppLibErrorPinMatchesVerificationCode, message: "New PIN must be different from verification code.")
    }

    public static func pinMatchesOldCodeError() -> NSError {
        return error(code: .moppLibErrorPinMatchesOldCode, message: "New PIN must be different from old PIN.")
    }

    public static func incorrectPinLengthError() -> NSError {
        return error(code: .moppLibErrorIncorrectPinLength, message: "PIN length didn't pass validation. Make sure minimum and maximum length requirements are met.")
    }

    public static func tooEasyPinError() -> NSError {
        return error(code: .moppLibErrorPinTooEasy, message: "New PIN code is too easy.")
    }

    public static func pinContainsInvalidCharactersError() -> NSError {
        return error(code: .moppLibErrorPinContainsInvalidCharacters, message: "New PIN contains invalid characters.")
    }

    public static func fileNameTooLongError() -> NSError {
        return error(code: .moppLibErrorFileNameTooLong, message: "File name is too long")
    }

    public static func noInternetConnectionError() -> NSError {
        return error(code: .moppLibErrorNoInternetConnection, message: "Internet connection not detected.")
    }

    public static func sslHandshakeError() -> NSError {
        return error(code: .moppLibErrorSslHandshakeFailed, message: "Failed to create SSL connection with host.")
    }

    public static func restrictedAPIError() -> NSError {
        return error(code: .moppLibErrorRestrictedApi, message: "This API method is not supported on third-party applications.")
    }

    public static func duplicatedFilenameError() -> NSError {
        return error(code: .moppLibErrorDuplicatedFilename, message: "Filename already exists")
    }

    public static func tooManyRequests() -> NSError {
        return error(code: .moppLibErrorTooManyRequests, message: "digidoc-service-error-too-many-requests")
    }

    public static func ocspTimeSlotError() -> NSError {
        return error(code: .moppLibErrorOCSPTimeSlot, message: "Invalid OCSP time slot")
    }

    public static func invalidProxySettingsError() -> NSError {
        return error(code: .moppLibErrorInvalidProxySettings, message: "Invalid proxy settings")
    }

    /// Generic function to create an `NSError` with a message
    private static func error(code: MoppLibErrorCode, message: String) -> NSError {
        return error(code: code, userInfo: [NSLocalizedDescriptionKey: message])
    }

    /// Generic function to create an `NSError` with custom `userInfo`
    private static func error(code: MoppLibErrorCode, userInfo: [String: Any]?) -> NSError {
        return NSError(domain: MoppLibErrorDomain, code: code.rawValue, userInfo: userInfo)
    }
}
